// crsf/dist/esm/index.mjs
/**
 * All possible values are precomputed
 * Uses CRC-8-Maxim variant
 * */
const CRC8_TABLE = [
    0x00, 0xd5, 0x7f, 0xaa, 0xfe, 0x2b, 0x81, 0x54, 0x29, 0xfc, 0x56, 0x83, 0xd7, 0x02, 0xa8, 0x7d,
    0x52, 0x87, 0x2d, 0xf8, 0xac, 0x79, 0xd3, 0x06, 0x7b, 0xae, 0x04, 0xd1, 0x85, 0x50, 0xfa, 0x2f,
    0xa4, 0x71, 0xdb, 0x0e, 0x5a, 0x8f, 0x25, 0xf0, 0x8d, 0x58, 0xf2, 0x27, 0x73, 0xa6, 0x0c, 0xd9,
    0xf6, 0x23, 0x89, 0x5c, 0x08, 0xdd, 0x77, 0xa2, 0xdf, 0x0a, 0xa0, 0x75, 0x21, 0xf4, 0x5e, 0x8b,
    0x9d, 0x48, 0xe2, 0x37, 0x63, 0xb6, 0x1c, 0xc9, 0xb4, 0x61, 0xcb, 0x1e, 0x4a, 0x9f, 0x35, 0xe0,
    0xcf, 0x1a, 0xb0, 0x65, 0x31, 0xe4, 0x4e, 0x9b, 0xe6, 0x33, 0x99, 0x4c, 0x18, 0xcd, 0x67, 0xb2,
    0x39, 0xec, 0x46, 0x93, 0xc7, 0x12, 0xb8, 0x6d, 0x10, 0xc5, 0x6f, 0xba, 0xee, 0x3b, 0x91, 0x44,
    0x6b, 0xbe, 0x14, 0xc1, 0x95, 0x40, 0xea, 0x3f, 0x42, 0x97, 0x3d, 0xe8, 0xbc, 0x69, 0xc3, 0x16,
    0xef, 0x3a, 0x90, 0x45, 0x11, 0xc4, 0x6e, 0xbb, 0xc6, 0x13, 0xb9, 0x6c, 0x38, 0xed, 0x47, 0x92,
    0xbd, 0x68, 0xc2, 0x17, 0x43, 0x96, 0x3c, 0xe9, 0x94, 0x41, 0xeb, 0x3e, 0x6a, 0xbf, 0x15, 0xc0,
    0x4b, 0x9e, 0x34, 0xe1, 0xb5, 0x60, 0xca, 0x1f, 0x62, 0xb7, 0x1d, 0xc8, 0x9c, 0x49, 0xe3, 0x36,
    0x19, 0xcc, 0x66, 0xb3, 0xe7, 0x32, 0x98, 0x4d, 0x30, 0xe5, 0x4f, 0x9a, 0xce, 0x1b, 0xb1, 0x64,
    0x72, 0xa7, 0x0d, 0xd8, 0x8c, 0x59, 0xf3, 0x26, 0x5b, 0x8e, 0x24, 0xf1, 0xa5, 0x70, 0xda, 0x0f,
    0x20, 0xf5, 0x5f, 0x8a, 0xde, 0x0b, 0xa1, 0x74, 0x09, 0xdc, 0x76, 0xa3, 0xf7, 0x22, 0x88, 0x5d,
    0xd6, 0x03, 0xa9, 0x7c, 0x28, 0xfd, 0x57, 0x82, 0xff, 0x2a, 0x80, 0x55, 0x01, 0xd4, 0x7e, 0xab,
    0x84, 0x51, 0xfb, 0x2e, 0x7a, 0xaf, 0x05, 0xd0, 0xad, 0x78, 0xd2, 0x07, 0x53, 0x86, 0x2c, 0xf9,
];
/**
 * Uses precomputed CRC8 table (CRC-8-Maxim variant)
 * @returns single byte value
 * */
const getCRC = (data) => data.reduce((crc, b) => CRC8_TABLE[crc ^ b], 0);

const DYNAMIC_ADDRESS_SPACE_FOR_NAT_START = 0x20;
const DYNAMIC_ADDRESS_SPACE_FOR_NAT_END = 0x7f;
const DYNAMIC_NAT_MAP = Object.fromEntries([
    ...Array((DYNAMIC_ADDRESS_SPACE_FOR_NAT_END -
        DYNAMIC_ADDRESS_SPACE_FOR_NAT_START +
        1)),
].map((_, i) => [`DYNAMIC_NAT_ADDRESS_${i}`, DYNAMIC_ADDRESS_SPACE_FOR_NAT_START + i]));
const SERIAL_SYNC_BYTE = 0xc8;
const DEVICE_ADDRESS = {
    BROADCAST_ADDRESS: 0x00,
    CLOUD: 0x0e,
    USB_DEVICE: 0x10,
    BLUETOOTH_MODULE_WIFI: 0x12,
    WIFI_RECEIVER_MOBILE_GAME_OR_SIMULATOR: 0x13,
    VIDEO_RECEIVER: 0x14,
    ...DYNAMIC_NAT_MAP,
    OSD_TBS_CORE_PNP_PRO: 0x80,
    ESC_1: 0x90,
    ESC_2: 0x91,
    ESC_3: 0x92,
    ESC_4: 0x93,
    ESC_5: 0x94,
    ESC_6: 0x95,
    ESC_7: 0x96,
    ESC_8: 0x97,
    RESERVED_0: 0x8a,
    CROSSFIRE_RESERVED_0: 0xb0,
    CROSSFIRE_RESERVED_1: 0xb2,
    VOLTAGE_OR_CURRENT_SENSOR: 0xc0,
    GPS_PNP_PRO_GPS: 0xc2,
    TBS_BLACK_BOX: 0xc4,
    FLIGHT_CONTROLLER: 0xc8,
    RESERVED_1: 0xca,
    RACE_TAG: 0xcc,
    VTX: 0xce,
    REMOTE_CONTROL: 0xea,
    RC_RECEIVER_CROSSFIRE_RX: 0xec,
    RC_TRANSMITTER_MODULE_CROSSFIRE_TX: 0xee,
    RESERVED_2: 0xf0,
    RESERVED_3: 0xf2,
};
const FRAME_TYPE = {
    UNSUPPORTED_OR_INVALID: 0x00, // used for invalid or unknown frame types
    GPS: 0x02,
    GPS_TIME: 0x03,
    GPS_EXTENDED: 0x06,
    VARIOMETER_SENSOR: 0x07,
    BATTERY_SENSOR: 0x08,
    BAROMETRIC_ALTITUDE_VERTICAL_SPEED: 0x09,
    AIRSPEED: 0x0a,
    HEARTBEAT: 0x0b,
    RPM: 0x0c,
    TEMP: 0x0d,
    VOLTAGES: 0x0e,
    DISCONTINUED: 0x0f, // VTX in ardupilot
    VTX_TELEMETRY: 0x10,
    BAROMETER: 0x11,
    MAGNETOMETER: 0x12,
    ACCEL_GYRO: 0x13,
    LINK_STATISTICS: 0x14,
    RC_CHANNELS_PACKED: 0x16,
    SUBSET_RC_CHANNELS_PACKED: 0x17,
    RC_CHANNELS_PACKED_11_BITS: 0x18,
    RESERVED_CROSSFIRE_0: 0x19,
    RESERVED_CROSSFIRE_1: 0x1a,
    RESERVED_CROSSFIRE_2: 0x1b,
    LINK_STATISTICS_RX: 0x1c,
    LINK_STATISTICS_TX: 0x1d,
    ATTITUDE: 0x1e,
    MAV_LINK_FC: 0x1f,
    FLIGHT_MODE: 0x21,
    ESP_NOW_MESSAGE: 0x22,
    RESERVED: 0x27,
    // extended
    PARAMETER_PING_DEVICES: 0x28,
    PARAMETER_DEVICE_INFORMATION: 0x29,
    // found in ardupilot
    PARAMETER_SETTINGS_ENTRY: 0x2b,
    PARAMETER_READ: 0x2c,
    PARAMETER_WRITE: 0x2d,
    FRAMETYPE_COMMAND: 0x32,
    AP_CUSTOM_TELEMETRY_LEGACY: 0x7f,
    AP_CUSTOM_TELEMETRY: 0x80,
};
const SYNC_BYTE_VARIANTS = new Set([SERIAL_SYNC_BYTE, ...Object.values(DEVICE_ADDRESS)]);
const MIN_FRAME_SIZE = 2; // type + crc
const MAX_FRAME_SIZE = 62; // per protocol specification
const MAX_PAYLOAD_SIZE = MAX_FRAME_SIZE - MIN_FRAME_SIZE;

class CrossfireFrame {
    constructor(syncByte = SERIAL_SYNC_BYTE, type = 0, payload = new Uint8Array(0)) {
        this.syncByte = syncByte;
        this.type = type;
        this.payload = payload;
    }
    /** Total frame length: type + payload + crc */
    get length() {
        return this.payload.length + 2;
    }
    get crc() {
        const crcData = new Uint8Array(1 + this.payload.length);
        crcData[0] = this.type;
        crcData.set(this.payload, 1);
        return getCRC(crcData);
    }
}

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise, SuppressedError, Symbol, Iterator */


function __esDecorate(ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
}
function __runInitializers(thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
}
function __setFunctionName(f, name, prefix) {
    if (typeof name === "symbol") name = name.description ? "[".concat(name.description, "]") : "";
    return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
}
function __classPrivateFieldGet(receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
}

function __classPrivateFieldSet(receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
}

typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
    var e = new Error(message);
    return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
};

/** Workaround for static interface implementation */
function staticImplements() {
    return (constructor) => {
    };
}
const UINT24_MAX = 0xffffff;
const setUint24BE = (view, offset, value) => {
    const clampedValue = Math.max(0, Math.min(UINT24_MAX, Math.trunc(value)));
    view.setUint8(offset, (clampedValue >> 16) & 0xff);
    view.setUint8(offset + 1, (clampedValue >> 8) & 0xff);
    view.setUint8(offset + 2, clampedValue & 0xff);
};
const getUint24BE = (view, offset) => (view.getUint8(offset) << 16) | (view.getUint8(offset + 1) << 8) | view.getUint8(offset + 2);
const INT24_MIN = -8388608;
const INT24_MAX = 0x7fffff;
const setInt24BE = (view, offset, value) => {
    const clamped = Math.max(INT24_MIN, Math.min(INT24_MAX, Math.trunc(value)));
    // Convert to unsigned 24-bit representation
    const v = clamped < 0 ? clamped + 0x1000000 : clamped;
    view.setUint8(offset, (v >> 16) & 0xff);
    view.setUint8(offset + 1, (v >> 8) & 0xff);
    view.setUint8(offset + 2, v & 0xff);
};
const getInt24BE = (view, offset) => {
    const v = (view.getUint8(offset) << 16) | (view.getUint8(offset + 1) << 8) | view.getUint8(offset + 2);
    // Sign bit check (bit 23)
    return v & 0x800000 ? v - 0x1000000 : v;
};
const encodeCString = (str) => {
    const encoder = new TextEncoder(); // UTF-8 but ASCII-safe
    const bytes = encoder.encode(str);
    const out = new Uint8Array(bytes.length + 1);
    out.set(bytes, 0);
    out[bytes.length] = 0x00;
    return out;
};
const decodeCString = (buffer) => {
    const end = buffer.indexOf(0x00);
    const slice = end === -1 ? buffer : buffer.subarray(0, end);
    return new TextDecoder().decode(slice);
};

let UnsupportedOrInvalid = (() => {
    var _UnsupportedOrInvalid_frameType, _UnsupportedOrInvalid_buffer;
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var UnsupportedOrInvalid = _classThis = class {
        constructor(frameType = FRAME_TYPE.UNSUPPORTED_OR_INVALID, buffer = new Uint8Array(0), syncByte = SERIAL_SYNC_BYTE) {
            this.syncByte = syncByte;
            _UnsupportedOrInvalid_frameType.set(this, void 0);
            _UnsupportedOrInvalid_buffer.set(this, void 0);
            __classPrivateFieldSet(this, _UnsupportedOrInvalid_frameType, frameType, "f");
            __classPrivateFieldSet(this, _UnsupportedOrInvalid_buffer, buffer, "f");
        }
        static get payloadSize() {
            return 0;
        }
        get frameType() {
            return __classPrivateFieldGet(this, _UnsupportedOrInvalid_frameType, "f");
        }
        get crossfireFrame() {
            return new CrossfireFrame(this.syncByte, __classPrivateFieldGet(this, _UnsupportedOrInvalid_frameType, "f"), __classPrivateFieldGet(this, _UnsupportedOrInvalid_buffer, "f"));
        }
        static fromFrame(frame) {
            return new UnsupportedOrInvalid(frame.type, frame.payload, frame.syncByte);
        }
    };
    _UnsupportedOrInvalid_frameType = new WeakMap();
    _UnsupportedOrInvalid_buffer = new WeakMap();
    __setFunctionName(_classThis, "UnsupportedOrInvalid");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        UnsupportedOrInvalid = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return UnsupportedOrInvalid = _classThis;
})();

let GPS = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var GPS = _classThis = class {
        constructor(
        /** int32 - degree / 10`000`000 */
        latitude = 0, 
        /** int32 - degree / 10`000`000 */
        longitude = 0, 
        /** uint16 - km/h / 100 */
        groundSpeed = 0, 
        /** uint16 - degree / 100 */
        heading = 0, 
        /** uint16 - meter - 1000m offset */
        altitude = 0, 
        /** uint8 - # of sats in view */
        satellites = 0, syncByte = SERIAL_SYNC_BYTE) {
            this.latitude = latitude;
            this.longitude = longitude;
            this.groundSpeed = groundSpeed;
            this.heading = heading;
            this.altitude = altitude;
            this.satellites = satellites;
            this.syncByte = syncByte;
        }
        static get payloadSize() {
            return (4 + // latitude int32
                4 + // longitude int32
                2 + // groundSpeed uint16
                2 + // heading uint16
                2 + // altitude uint16
                1 // satellites uint8
            );
        }
        get frameType() {
            return FRAME_TYPE.GPS;
        }
        get crossfireFrame() {
            const payload = new Uint8Array(GPS.payloadSize);
            const payloadView = new DataView(payload.buffer);
            let offset = 0;
            payloadView.setInt32(offset, this.latitude);
            offset += 4;
            payloadView.setInt32(offset, this.longitude);
            offset += 4;
            payloadView.setUint16(offset, this.groundSpeed);
            offset += 2;
            payloadView.setUint16(offset, this.heading);
            offset += 2;
            payloadView.setUint16(offset, this.altitude);
            offset += 2;
            payloadView.setUint8(offset, this.satellites);
            offset += 1;
            return new CrossfireFrame(this.syncByte, this.frameType, new Uint8Array(payloadView.buffer));
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.GPS)
                throw new Error(`Invalid frame type for GPS: ${frame.type.toString(16)}`);
            if (frame.payload.length < GPS.payloadSize)
                return UnsupportedOrInvalid.fromFrame(frame);
            const payloadView = new DataView(frame.payload.buffer, frame.payload.byteOffset, frame.payload.byteLength);
            let offset = 0;
            const latitude = payloadView.getInt32(offset);
            offset += 4;
            const longitude = payloadView.getInt32(offset);
            offset += 4;
            const groundSpeed = payloadView.getUint16(offset);
            offset += 2;
            const heading = payloadView.getUint16(offset);
            offset += 2;
            const altitude = payloadView.getUint16(offset);
            offset += 2;
            const satellites = payloadView.getUint8(offset);
            offset += 1;
            return new GPS(latitude, longitude, groundSpeed, heading, altitude, satellites, frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "GPS");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        GPS = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return GPS = _classThis;
})();

let GPSTime = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var GPSTime = _classThis = class {
        constructor(
        /** uint16 */
        year = 2000, 
        /** uint8 */
        month = 1, 
        /** uint8 */
        day = 1, 
        /** uint8 */
        hour = 0, 
        /** uint8 */
        minute = 0, 
        /** uint8 */
        second = 0, 
        /** uint16 */
        millisecond = 0, syncByte = SERIAL_SYNC_BYTE) {
            this.year = year;
            this.month = month;
            this.day = day;
            this.hour = hour;
            this.minute = minute;
            this.second = second;
            this.millisecond = millisecond;
            this.syncByte = syncByte;
        }
        static get payloadSize() {
            return (2 + // year uint16
                1 + // month uint8
                1 + // day uint8
                1 + // hour uint8
                1 + // minute uint8
                1 + // second uint8
                2 // millisecond uint16
            );
        }
        get frameType() {
            return FRAME_TYPE.GPS_TIME;
        }
        get crossfireFrame() {
            const payload = new Uint8Array(GPSTime.payloadSize);
            const payloadView = new DataView(payload.buffer);
            let offset = 0;
            payloadView.setUint16(offset, this.year);
            offset += 2;
            payloadView.setUint8(offset, this.month);
            offset += 1;
            payloadView.setUint8(offset, this.day);
            offset += 1;
            payloadView.setUint8(offset, this.hour);
            offset += 1;
            payloadView.setUint8(offset, this.minute);
            offset += 1;
            payloadView.setUint8(offset, this.second);
            offset += 1;
            payloadView.setUint16(offset, this.millisecond);
            offset += 2;
            return new CrossfireFrame(this.syncByte, this.frameType, new Uint8Array(payloadView.buffer));
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.GPS_TIME)
                throw new Error(`Invalid frame type for GPSTime: ${frame.type.toString(16)}`);
            if (frame.payload.length < GPSTime.payloadSize)
                return UnsupportedOrInvalid.fromFrame(frame);
            const payloadView = new DataView(frame.payload.buffer, frame.payload.byteOffset, frame.payload.byteLength);
            let offset = 0;
            const year = payloadView.getUint16(offset);
            offset += 2;
            const month = payloadView.getUint8(offset);
            offset += 1;
            const day = payloadView.getUint8(offset);
            offset += 1;
            const hour = payloadView.getUint8(offset);
            offset += 1;
            const minute = payloadView.getUint8(offset);
            offset += 1;
            const second = payloadView.getUint8(offset);
            offset += 1;
            const millisecond = payloadView.getUint16(offset);
            offset += 2;
            return new GPSTime(year, month, day, hour, minute, second, millisecond, frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "GPSTime");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        GPSTime = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return GPSTime = _classThis;
})();

let GPSExtended = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var GPSExtended = _classThis = class {
        constructor(
        /** uint8 - Current GPS fix quality */
        fixType = 0, 
        /** int16 - Northward (north = positive) Speed [cm/sec] */
        nSpeed = 0, 
        /** int16 - Eastward (east = positive) Speed [cm/sec] */
        eSpeed = 0, 
        /** int16 - Vertical (up = positive) Speed [cm/sec] */
        vSpeed = 0, 
        /** int16 - Horizontal Speed accuracy cm/sec */
        hSpeedAcc = 0, 
        /** int16 - Heading accuracy in degrees scaled with 1e-1 degrees times 10) */
        trackAcc = 0, 
        /** int16 - Meters Height above GPS Ellipsoid (not MSL) */
        altEllipsoid = 0, 
        /** int16 - horizontal accuracy in cm */
        hAcc = 0, 
        /** int16 - vertical accuracy in cm */
        vAcc = 0, 
        /** uint8 - reserved */
        reserved = 0, 
        /** uint8 - horizontal dilution of precision,Dimensionless in nits of.1. */
        hDOP = 0, 
        /** uint8 - vertical dilution of precision, Dimensionless in nits of .1. */
        vDOP = 0, syncByte = SERIAL_SYNC_BYTE) {
            this.fixType = fixType;
            this.nSpeed = nSpeed;
            this.eSpeed = eSpeed;
            this.vSpeed = vSpeed;
            this.hSpeedAcc = hSpeedAcc;
            this.trackAcc = trackAcc;
            this.altEllipsoid = altEllipsoid;
            this.hAcc = hAcc;
            this.vAcc = vAcc;
            this.reserved = reserved;
            this.hDOP = hDOP;
            this.vDOP = vDOP;
            this.syncByte = syncByte;
        }
        static get payloadSize() {
            return (1 + // fixType uint8
                2 + // nSpeed int16
                2 + // eSpeed int16
                2 + // vSpeed int16
                2 + // hSpeedAcc int16
                2 + // trackAcc int16
                2 + // altEllipsoid int16
                2 + // hAcc int16
                2 + // vAcc int16
                1 + // reserved uint8
                1 + // hDOP uint8
                1 // vDOP uint8
            );
        }
        get frameType() {
            return FRAME_TYPE.GPS_EXTENDED;
        }
        get crossfireFrame() {
            const payload = new Uint8Array(GPSExtended.payloadSize);
            const payloadView = new DataView(payload.buffer);
            let offset = 0;
            payloadView.setUint8(offset, this.fixType);
            offset += 1;
            payloadView.setInt16(offset, this.nSpeed);
            offset += 2;
            payloadView.setInt16(offset, this.eSpeed);
            offset += 2;
            payloadView.setInt16(offset, this.vSpeed);
            offset += 2;
            payloadView.setInt16(offset, this.hSpeedAcc);
            offset += 2;
            payloadView.setInt16(offset, this.trackAcc);
            offset += 2;
            payloadView.setInt16(offset, this.altEllipsoid);
            offset += 2;
            payloadView.setInt16(offset, this.hAcc);
            offset += 2;
            payloadView.setInt16(offset, this.vAcc);
            offset += 2;
            payloadView.setUint8(offset, this.reserved);
            offset += 1;
            payloadView.setUint8(offset, this.hDOP);
            offset += 1;
            payloadView.setUint8(offset, this.vDOP);
            offset += 1;
            return new CrossfireFrame(this.syncByte, this.frameType, new Uint8Array(payloadView.buffer));
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.GPS_EXTENDED)
                throw new Error(`Invalid frame type for GPS Extended: ${frame.type.toString(16)}`);
            if (frame.payload.length < GPSExtended.payloadSize)
                return UnsupportedOrInvalid.fromFrame(frame);
            const payloadView = new DataView(frame.payload.buffer, frame.payload.byteOffset, frame.payload.byteLength);
            let offset = 0;
            const fixType = payloadView.getUint8(offset);
            offset += 1;
            const nSpeed = payloadView.getInt16(offset);
            offset += 2;
            const eSpeed = payloadView.getInt16(offset);
            offset += 2;
            const vSpeed = payloadView.getInt16(offset);
            offset += 2;
            const hSpeedAcc = payloadView.getInt16(offset);
            offset += 2;
            const trackAcc = payloadView.getInt16(offset);
            offset += 2;
            const altEllipsoid = payloadView.getInt16(offset);
            offset += 2;
            const hAcc = payloadView.getInt16(offset);
            offset += 2;
            const vAcc = payloadView.getInt16(offset);
            offset += 2;
            const reserved = payloadView.getUint8(offset);
            offset += 1;
            const hDOP = payloadView.getUint8(offset);
            offset += 1;
            const vDOP = payloadView.getUint8(offset);
            offset += 1;
            return new GPSExtended(fixType, nSpeed, eSpeed, vSpeed, hSpeedAcc, trackAcc, altEllipsoid, hAcc, vAcc, reserved, hDOP, vDOP, frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "GPSExtended");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        GPSExtended = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return GPSExtended = _classThis;
})();

let VariometerSensor = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var VariometerSensor = _classThis = class {
        constructor(
        /** int16 - Vertical speed cm/s */
        vSpeed = 0, syncByte = SERIAL_SYNC_BYTE) {
            this.vSpeed = vSpeed;
            this.syncByte = syncByte;
        }
        static get payloadSize() {
            return 2; // vSpeed int16
        }
        get frameType() {
            return FRAME_TYPE.VARIOMETER_SENSOR;
        }
        get crossfireFrame() {
            const payload = new Uint8Array(VariometerSensor.payloadSize);
            const payloadView = new DataView(payload.buffer);
            let offset = 0;
            payloadView.setInt16(offset, this.vSpeed);
            offset += 2;
            return new CrossfireFrame(this.syncByte, this.frameType, new Uint8Array(payloadView.buffer));
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.VARIOMETER_SENSOR) {
                throw new Error(`Invalid frame type for VariometerSensor: ${frame.type.toString(16)}`);
            }
            if (frame.payload.length < VariometerSensor.payloadSize) {
                return new UnsupportedOrInvalid(frame.type, frame.payload, frame.syncByte);
            }
            const payloadView = new DataView(frame.payload.buffer, frame.payload.byteOffset, frame.payload.byteLength);
            const vSpeed = payloadView.getInt16(0);
            return new VariometerSensor(vSpeed, frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "VariometerSensor");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        VariometerSensor = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return VariometerSensor = _classThis;
})();

let BatterySensor = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var BatterySensor = _classThis = class {
        constructor(
        /** int16 - Voltage (LSB = 10 µV) */
        voltage = 0, 
        /** int16 - Current (LSB = 10 µA) */
        current = 0, 
        /** uint24 - Capacity used (mAh) */
        capacityUsed = 0, 
        /** uint8 - Battery remaining (percent) */
        remaining = 0, syncByte = SERIAL_SYNC_BYTE) {
            this.voltage = voltage;
            this.current = current;
            this.capacityUsed = capacityUsed;
            this.remaining = remaining;
            this.syncByte = syncByte;
        }
        static get payloadSize() {
            return (2 + // voltage int16
                2 + // current int16
                3 + // capacityUsed uint24
                1 // remaining uint8
            );
        }
        get frameType() {
            return FRAME_TYPE.BATTERY_SENSOR;
        }
        get crossfireFrame() {
            const payload = new Uint8Array(BatterySensor.payloadSize);
            const payloadView = new DataView(payload.buffer);
            let offset = 0;
            payloadView.setInt16(offset, this.voltage);
            offset += 2;
            payloadView.setInt16(offset, this.current);
            offset += 2;
            setUint24BE(payloadView, offset, this.capacityUsed);
            offset += 3;
            payloadView.setUint8(offset, this.remaining);
            offset += 1;
            return new CrossfireFrame(this.syncByte, this.frameType, new Uint8Array(payloadView.buffer));
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.BATTERY_SENSOR) {
                throw new Error(`Invalid frame type for BatterySensor: ${frame.type.toString(16)}`);
            }
            if (frame.payload.length < BatterySensor.payloadSize) {
                return new UnsupportedOrInvalid(frame.type, frame.payload, frame.syncByte);
            }
            const payloadView = new DataView(frame.payload.buffer, frame.payload.byteOffset, frame.payload.byteLength);
            let offset = 0;
            const voltage = payloadView.getInt16(offset);
            offset += 2;
            const current = payloadView.getInt16(offset);
            offset += 2;
            const capacityUsed = getUint24BE(payloadView, offset);
            offset += 3;
            const remaining = payloadView.getUint8(offset);
            offset += 1;
            return new BatterySensor(voltage, current, capacityUsed, remaining, frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "BatterySensor");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        BatterySensor = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return BatterySensor = _classThis;
})();

let BarometricAltitudeVerticalSpeed = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var BarometricAltitudeVerticalSpeed = _classThis = class {
        constructor(
        /** uint16 - Altitude above start (calibration) point */
        altitudePacked = 0, 
        /** int8 - vertical speed */
        verticalSpeedPacked = 0, syncByte = SERIAL_SYNC_BYTE) {
            this.altitudePacked = altitudePacked;
            this.verticalSpeedPacked = verticalSpeedPacked;
            this.syncByte = syncByte;
        }
        static get payloadSize() {
            return (2 + // altitudePacked uint16
                1 // verticalSpeedPacked int8
            );
        }
        get frameType() {
            return FRAME_TYPE.BAROMETRIC_ALTITUDE_VERTICAL_SPEED;
        }
        get crossfireFrame() {
            const payload = new Uint8Array(BarometricAltitudeVerticalSpeed.payloadSize);
            const payloadView = new DataView(payload.buffer);
            let offset = 0;
            payloadView.setUint16(offset, this.altitudePacked);
            offset += 2;
            payloadView.setInt8(offset, this.verticalSpeedPacked);
            offset += 1;
            return new CrossfireFrame(this.syncByte, this.frameType, new Uint8Array(payloadView.buffer));
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.BAROMETRIC_ALTITUDE_VERTICAL_SPEED) {
                throw new Error(`Invalid frame type for BarometricAltitudeVerticalSpeed: ${frame.type.toString(16)}`);
            }
            if (frame.payload.length < BarometricAltitudeVerticalSpeed.payloadSize) {
                return new UnsupportedOrInvalid(frame.type, frame.payload, frame.syncByte);
            }
            const payloadView = new DataView(frame.payload.buffer, frame.payload.byteOffset, frame.payload.byteLength);
            let offset = 0;
            const altitudePacked = payloadView.getUint16(offset);
            offset += 2;
            const verticalSpeedPacked = payloadView.getInt8(offset);
            offset += 1;
            return new BarometricAltitudeVerticalSpeed(altitudePacked, verticalSpeedPacked, frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "BarometricAltitudeVerticalSpeed");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        BarometricAltitudeVerticalSpeed = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return BarometricAltitudeVerticalSpeed = _classThis;
})();

let Airspeed = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var Airspeed = _classThis = class {
        constructor(
        /** uint16 - Airspeed in 0.1 * km/h (hectometers/h) */
        speed = 0, syncByte = SERIAL_SYNC_BYTE) {
            this.speed = speed;
            this.syncByte = syncByte;
        }
        static get payloadSize() {
            return 2; // speed uint16
        }
        get frameType() {
            return FRAME_TYPE.AIRSPEED;
        }
        get crossfireFrame() {
            const payload = new Uint8Array(Airspeed.payloadSize);
            const payloadView = new DataView(payload.buffer);
            payloadView.setUint16(0, this.speed);
            return new CrossfireFrame(this.syncByte, this.frameType, payload);
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.AIRSPEED) {
                throw new Error(`Invalid frame type for Airspeed: ${frame.type.toString(16)}`);
            }
            if (frame.payload.length < Airspeed.payloadSize) {
                return new UnsupportedOrInvalid(frame.type, frame.payload, frame.syncByte);
            }
            const payloadView = new DataView(frame.payload.buffer, frame.payload.byteOffset, frame.payload.byteLength);
            const speed = payloadView.getUint16(0);
            return new Airspeed(speed, frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "Airspeed");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        Airspeed = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return Airspeed = _classThis;
})();

let Heartbeat = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var Heartbeat = _classThis = class {
        constructor(
        /** int16 - Origin Device address */
        originAddress = 0, syncByte = SERIAL_SYNC_BYTE) {
            this.originAddress = originAddress;
            this.syncByte = syncByte;
        }
        static get payloadSize() {
            return 2; // originAddress int16
        }
        get frameType() {
            return FRAME_TYPE.HEARTBEAT;
        }
        get crossfireFrame() {
            const payload = new Uint8Array(Heartbeat.payloadSize);
            const payloadView = new DataView(payload.buffer);
            payloadView.setInt16(0, this.originAddress);
            return new CrossfireFrame(this.syncByte, this.frameType, new Uint8Array(payloadView.buffer));
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.HEARTBEAT) {
                throw new Error(`Invalid frame type for Heartbeat: ${frame.type.toString(16)}`);
            }
            if (frame.payload.length < Heartbeat.payloadSize) {
                return new UnsupportedOrInvalid(frame.type, frame.payload, frame.syncByte);
            }
            const payloadView = new DataView(frame.payload.buffer, frame.payload.byteOffset, frame.payload.byteLength);
            const originAddress = payloadView.getInt16(0);
            return new Heartbeat(originAddress, frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "Heartbeat");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        Heartbeat = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return Heartbeat = _classThis;
})();

let RMP = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var RMP = _classThis = class {
        constructor(
        /** uint8 - Identifies the source of the RPM data (e.g., 0 = Motor 1, 1 = Motor 2, etc.) */
        rpmSourceId = 0, 
        /** int24[] - 1 - 19 RPM values with negative ones representing the motor spinning in reverse */
        rpmValues = new Array(19).fill(0), syncByte = SERIAL_SYNC_BYTE) {
            this.rpmSourceId = rpmSourceId;
            this.rpmValues = rpmValues;
            this.syncByte = syncByte;
            if (this.rpmValues.length !== 19) {
                throw new Error("rpmValues must have exactly 19 elements");
            }
        }
        static get payloadSize() {
            return (1 + // rpmSourceId uint8
                3 * 19 // rpmValues int24[]
            );
        }
        get frameType() {
            return FRAME_TYPE.RPM;
        }
        get crossfireFrame() {
            if (this.rpmValues.length !== 19) {
                throw new Error("rpmValues must have exactly 19 elements");
            }
            const payload = new Uint8Array(RMP.payloadSize);
            const payloadView = new DataView(payload.buffer);
            let offset = 0;
            payloadView.setUint8(offset, this.rpmSourceId);
            offset += 1;
            for (const rpmValue of this.rpmValues) {
                setInt24BE(payloadView, offset, rpmValue);
                offset += 3;
            }
            return new CrossfireFrame(this.syncByte, this.frameType, new Uint8Array(payloadView.buffer));
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.RPM) {
                throw new Error(`Invalid frame type for RMP: ${frame.type.toString(16)}`);
            }
            if (frame.payload.length < RMP.payloadSize) {
                return new UnsupportedOrInvalid(frame.type, frame.payload, frame.syncByte);
            }
            const payloadView = new DataView(frame.payload.buffer, frame.payload.byteOffset, frame.payload.byteLength);
            let offset = 0;
            const rpmSourceId = payloadView.getUint8(offset);
            offset += 1;
            const rpmValues = new Array(19).fill(0);
            for (let i = 0; i < 19; i++) {
                const rpmValue = getInt24BE(payloadView, offset);
                rpmValues[i] = rpmValue;
                offset += 3;
            }
            return new RMP(rpmSourceId, rpmValues, frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "RMP");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        RMP = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return RMP = _classThis;
})();

let Temp = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var Temp = _classThis = class {
        constructor(
        /** uint8 - source of the temperature data (e.g., 0 = FC including all ESCs, 1 = Ambient, etc.) */
        tempSourceId = 0, 
        /** int16[] - up to 20 temperature values in deci-degree (tenths of a degree) Celsius (e.g., 250 = 25.0°C, -50 = -5.0°C) */
        temperatures = new Array(20).fill(0), syncByte = SERIAL_SYNC_BYTE) {
            this.tempSourceId = tempSourceId;
            this.temperatures = temperatures;
            this.syncByte = syncByte;
            if (this.temperatures.length !== 20) {
                throw new Error("temperatures must have exactly 20 elements");
            }
        }
        static get payloadSize() {
            return (1 + // tempSourceId uint8
                2 * 20 // temperatures int16[]
            );
        }
        get frameType() {
            return FRAME_TYPE.TEMP;
        }
        get crossfireFrame() {
            if (this.temperatures.length !== 20) {
                throw new Error("temperatures must have exactly 20 elements");
            }
            const payload = new Uint8Array(Temp.payloadSize);
            const payloadView = new DataView(payload.buffer);
            let offset = 0;
            payloadView.setUint8(offset, this.tempSourceId);
            offset += 1;
            for (const tempValue of this.temperatures) {
                payloadView.setInt16(offset, tempValue);
                offset += 2;
            }
            return new CrossfireFrame(this.syncByte, this.frameType, new Uint8Array(payloadView.buffer));
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.TEMP) {
                throw new Error(`Invalid frame type for Temp: ${frame.type.toString(16)}`);
            }
            if (frame.payload.length < Temp.payloadSize) {
                return new UnsupportedOrInvalid(frame.type, frame.payload, frame.syncByte);
            }
            const payloadView = new DataView(frame.payload.buffer, frame.payload.byteOffset, frame.payload.byteLength);
            let offset = 0;
            const tempSourceId = payloadView.getUint8(offset);
            offset += 1;
            const temperatures = new Array(20).fill(0);
            for (let i = 0; i < 20; i++) {
                temperatures[i] = payloadView.getInt16(offset);
                offset += 2;
            }
            return new Temp(tempSourceId, temperatures, frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "Temp");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        Temp = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return Temp = _classThis;
})();

let Voltages = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var Voltages = _classThis = class {
        constructor(
        /** uint8 - source of the voltages */
        sourceId = 0, 
        /** uint16[] - up to 29 voltages in millivolts (e.g. 3.850V = 3850) */
        voltages = new Array(29).fill(0), syncByte = SERIAL_SYNC_BYTE) {
            this.sourceId = sourceId;
            this.voltages = voltages;
            this.syncByte = syncByte;
            if (this.voltages.length !== 29) {
                throw new Error("voltages must have exactly 29 elements");
            }
        }
        static get payloadSize() {
            return (1 + // sourceId uint8
                2 * 29 // voltages uint16[]
            );
        }
        get frameType() {
            return FRAME_TYPE.VOLTAGES;
        }
        get crossfireFrame() {
            if (this.voltages.length !== 29)
                throw new Error("voltages must have exactly 29 elements");
            const payload = new Uint8Array(Voltages.payloadSize);
            const payloadView = new DataView(payload.buffer);
            let offset = 0;
            payloadView.setUint8(offset, this.sourceId);
            offset += 1;
            for (let i = 0; i < 29; i++) {
                const voltage = this.voltages[i] || 0;
                payloadView.setUint16(offset, voltage);
                offset += 2;
            }
            return new CrossfireFrame(this.syncByte, this.frameType, new Uint8Array(payloadView.buffer));
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.VOLTAGES) {
                throw new Error(`Invalid frame type for Voltages: ${frame.type.toString(16)}`);
            }
            if (frame.payload.length < Voltages.payloadSize) {
                return new UnsupportedOrInvalid(frame.type, frame.payload, frame.syncByte);
            }
            const payloadView = new DataView(frame.payload.buffer, frame.payload.byteOffset, frame.payload.byteLength);
            let offset = 0;
            const sourceId = payloadView.getUint8(offset);
            offset += 1;
            const voltages = [];
            for (let i = 0; i < 29; i++) {
                voltages.push(payloadView.getUint16(offset));
                offset += 2;
            }
            return new Voltages(sourceId, voltages, frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "Voltages");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        Voltages = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return Voltages = _classThis;
})();

let Discontinued = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var Discontinued = _classThis = class {
        constructor(syncByte = SERIAL_SYNC_BYTE) {
            this.syncByte = syncByte;
        }
        static get payloadSize() {
            return 0;
        }
        get frameType() {
            return FRAME_TYPE.DISCONTINUED;
        }
        get crossfireFrame() {
            const payload = new Uint8Array(Discontinued.payloadSize);
            return new CrossfireFrame(this.syncByte, this.frameType, payload);
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.DISCONTINUED) {
                throw new Error(`Invalid frame type for Di: ${frame.type.toString(16)}`);
            }
            return new Discontinued(frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "Discontinued");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        Discontinued = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return Discontinued = _classThis;
})();

let VTXTelemetry = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var VTXTelemetry = _classThis = class {
        constructor(
        /** uint8 */
        originAddress = 0, 
        /** uint8 - VTX power in dBm */
        powerDbm = 0, 
        /** uint16 - VTX frequency in MHz */
        frequencyMhz = 0, 
        /** uint8 - 0=Off, 1=On */
        pitMode = 1, 
        /** uint8 - 0=Off, 1=On, 2=Switch, 3=Failsafe */
        pitModeControl = 2, 
        /** uint8 - 0=Ch5, 1=Ch5 Inv, … , 15=Ch12 Inv */
        pitModeSwitch = 4, syncByte = SERIAL_SYNC_BYTE) {
            this.originAddress = originAddress;
            this.powerDbm = powerDbm;
            this.frequencyMhz = frequencyMhz;
            this.pitMode = pitMode;
            this.pitModeControl = pitModeControl;
            this.pitModeSwitch = pitModeSwitch;
            this.syncByte = syncByte;
        }
        static get payloadSize() {
            return (1 + // originAddress uint8
                1 + // powerDbm uint8
                2 + // frequencyMhz uint16
                1 + // pitMode uint8
                1 + // pitModeControl uint8
                1 // pitModeSwitch uint8
            );
        }
        get frameType() {
            return FRAME_TYPE.VTX_TELEMETRY;
        }
        get crossfireFrame() {
            const payload = new Uint8Array(VTXTelemetry.payloadSize);
            const payloadView = new DataView(payload.buffer);
            let offset = 0;
            payloadView.setUint8(offset, this.originAddress);
            offset += 1;
            payloadView.setUint8(offset, this.powerDbm);
            offset += 1;
            payloadView.setUint16(offset, this.frequencyMhz);
            offset += 2;
            payloadView.setUint8(offset, this.pitMode);
            offset += 1;
            payloadView.setUint8(offset, this.pitModeControl);
            offset += 1;
            payloadView.setUint8(offset, this.pitModeSwitch);
            offset += 1;
            return new CrossfireFrame(this.syncByte, this.frameType, new Uint8Array(payloadView.buffer));
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.VTX_TELEMETRY) {
                throw new Error(`Invalid frame type for VTXTelemetry: ${frame.type.toString(16)}`);
            }
            if (frame.payload.length < VTXTelemetry.payloadSize) {
                return new UnsupportedOrInvalid(frame.type, frame.payload, frame.syncByte);
            }
            const payloadView = new DataView(frame.payload.buffer, frame.payload.byteOffset, frame.payload.byteLength);
            let offset = 0;
            const originAddress = payloadView.getUint8(offset);
            offset += 1;
            const powerDbm = payloadView.getUint8(offset);
            offset += 1;
            const frequencyMhz = payloadView.getUint16(offset);
            offset += 2;
            const pitMode = payloadView.getUint8(offset);
            offset += 1;
            const pitModeControl = payloadView.getUint8(offset);
            offset += 1;
            const pitModeSwitch = payloadView.getUint8(offset);
            offset += 1;
            return new VTXTelemetry(originAddress, powerDbm, frequencyMhz, pitMode, pitModeControl, pitModeSwitch, frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "VTXTelemetry");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        VTXTelemetry = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return VTXTelemetry = _classThis;
})();

let Barometer = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var Barometer = _classThis = class {
        constructor(
        /** int32 - Pascals */
        pressurePa = 0, 
        /** int32 - centidegrees */
        baroTemp = 0, syncByte = SERIAL_SYNC_BYTE) {
            this.pressurePa = pressurePa;
            this.baroTemp = baroTemp;
            this.syncByte = syncByte;
        }
        static get payloadSize() {
            return (4 + // pressurePa int32
                4 // baroTemp int32
            );
        }
        get frameType() {
            return FRAME_TYPE.BAROMETER;
        }
        get crossfireFrame() {
            const payload = new Uint8Array(Barometer.payloadSize);
            const payloadView = new DataView(payload.buffer);
            let offset = 0;
            payloadView.setInt32(offset, this.pressurePa);
            offset += 4;
            payloadView.setInt32(offset, this.baroTemp);
            offset += 4;
            return new CrossfireFrame(this.syncByte, this.frameType, new Uint8Array(payloadView.buffer));
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.BAROMETER) {
                throw new Error(`Invalid frame type for Barometer: ${frame.type.toString(16)}`);
            }
            if (frame.payload.length < Barometer.payloadSize) {
                return new UnsupportedOrInvalid(frame.type, frame.payload, frame.syncByte);
            }
            const payloadView = new DataView(frame.payload.buffer, frame.payload.byteOffset, frame.payload.byteLength);
            let offset = 0;
            const pressurePa = payloadView.getInt32(offset);
            offset += 4;
            const baroTemp = payloadView.getInt32(offset);
            offset += 4;
            return new Barometer(pressurePa, baroTemp, frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "Barometer");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        Barometer = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return Barometer = _classThis;
})();

let Magnetometer = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var Magnetometer = _classThis = class {
        constructor(
        /** int16 - milligauss * 3 */
        fieldX = 0, 
        /** int16 - milligauss * 3 */
        fieldY = 0, 
        /** int16 - milligauss * 3 */
        fieldZ = 0, syncByte = SERIAL_SYNC_BYTE) {
            this.fieldX = fieldX;
            this.fieldY = fieldY;
            this.fieldZ = fieldZ;
            this.syncByte = syncByte;
        }
        static get payloadSize() {
            return (2 + // fieldX int16
                2 + // fieldY int16
                2 // fieldZ int16
            );
        }
        get frameType() {
            return FRAME_TYPE.MAGNETOMETER;
        }
        get crossfireFrame() {
            const payload = new Uint8Array(Magnetometer.payloadSize);
            const payloadView = new DataView(payload.buffer);
            let offset = 0;
            payloadView.setInt16(offset, this.fieldX);
            offset += 2;
            payloadView.setInt16(offset, this.fieldY);
            offset += 2;
            payloadView.setInt16(offset, this.fieldZ);
            offset += 2;
            return new CrossfireFrame(this.syncByte, this.frameType, new Uint8Array(payloadView.buffer));
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.MAGNETOMETER) {
                throw new Error(`Invalid frame type for Magnetometer: ${frame.type.toString(16)}`);
            }
            if (frame.payload.length < Magnetometer.payloadSize) {
                return new UnsupportedOrInvalid(frame.type, frame.payload, frame.syncByte);
            }
            const payloadView = new DataView(frame.payload.buffer, frame.payload.byteOffset, frame.payload.byteLength);
            let offset = 0;
            const fieldX = payloadView.getInt16(offset);
            offset += 2;
            const fieldY = payloadView.getInt16(offset);
            offset += 2;
            const fieldZ = payloadView.getInt16(offset);
            offset += 2;
            return new Magnetometer(fieldX, fieldY, fieldZ, frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "Magnetometer");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        Magnetometer = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return Magnetometer = _classThis;
})();

let AccelGyro = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var AccelGyro = _classThis = class {
        constructor(
        /** uint32 - Timestamp of the sample in us */
        sampleTime = 0, 
        /** int16 - SB = INT16_MAX/2000 DPS */
        gyroX = 0, 
        /** int16 - SB = INT16_MAX/2000 DPS */
        gyroY = 0, 
        /** int16 - SB = INT16_MAX/2000 DPS */
        gyroZ = 0, 
        /** int16 - LSB = INT16_MAX/16 G */
        accX = 0, 
        /** int16 - LSB = INT16_MAX/16 G */
        accY = 0, 
        /** int16 - LSB = INT16_MAX/16 G */
        accZ = 0, 
        /** int16 - centidegrees */
        gyroTemp = 0, syncByte = SERIAL_SYNC_BYTE) {
            this.sampleTime = sampleTime;
            this.gyroX = gyroX;
            this.gyroY = gyroY;
            this.gyroZ = gyroZ;
            this.accX = accX;
            this.accY = accY;
            this.accZ = accZ;
            this.gyroTemp = gyroTemp;
            this.syncByte = syncByte;
        }
        static get payloadSize() {
            return (4 + // sampleTime uint32
                2 + // gyroX int16
                2 + // gyroY int16
                2 + // gyroZ int16
                2 + // accX int16
                2 + // accY int16
                2 + // accZ int16
                2 // gyroTemp int16
            );
        }
        get frameType() {
            return FRAME_TYPE.ACCEL_GYRO;
        }
        get crossfireFrame() {
            const payload = new Uint8Array(AccelGyro.payloadSize);
            const payloadView = new DataView(payload.buffer);
            let offset = 0;
            payloadView.setUint32(offset, this.sampleTime);
            offset += 4;
            payloadView.setInt16(offset, this.gyroX);
            offset += 2;
            payloadView.setInt16(offset, this.gyroY);
            offset += 2;
            payloadView.setInt16(offset, this.gyroZ);
            offset += 2;
            payloadView.setInt16(offset, this.accX);
            offset += 2;
            payloadView.setInt16(offset, this.accY);
            offset += 2;
            payloadView.setInt16(offset, this.accZ);
            offset += 2;
            payloadView.setInt16(offset, this.gyroTemp);
            offset += 2;
            return new CrossfireFrame(this.syncByte, this.frameType, new Uint8Array(payloadView.buffer));
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.ACCEL_GYRO) {
                throw new Error(`Invalid frame type for AccelGyro: ${frame.type.toString(16)}`);
            }
            if (frame.payload.length < AccelGyro.payloadSize) {
                return new UnsupportedOrInvalid(frame.type, frame.payload, frame.syncByte);
            }
            const payloadView = new DataView(frame.payload.buffer, frame.payload.byteOffset, frame.payload.byteLength);
            let offset = 0;
            const sampleTime = payloadView.getUint32(offset);
            offset += 4;
            const gyroX = payloadView.getInt16(offset);
            offset += 2;
            const gyroY = payloadView.getInt16(offset);
            offset += 2;
            const gyroZ = payloadView.getInt16(offset);
            offset += 2;
            const accX = payloadView.getInt16(offset);
            offset += 2;
            const accY = payloadView.getInt16(offset);
            offset += 2;
            const accZ = payloadView.getInt16(offset);
            offset += 2;
            const gyroTemp = payloadView.getInt16(offset);
            offset += 2;
            return new AccelGyro(sampleTime, gyroX, gyroY, gyroZ, accX, accY, accZ, gyroTemp, frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "AccelGyro");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        AccelGyro = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return AccelGyro = _classThis;
})();

let LinkStatistics = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var LinkStatistics = _classThis = class {
        constructor(
        /** uint8 - Uplink RSSI Antenna 1 (dBm * -1) */
        puRssiAnt1 = 0, 
        /** uint8 - Uplink RSSI Antenna 2 (dBm * -1) */
        puRssiAnt2 = 0, 
        /** uint8 - Uplink Package success rate / Link quality (%) */
        upLinkQuality = 0, 
        /** int8 - Uplink SNR (dB) */
        upSnr = 0, 
        /** uint8 - number of currently best antenna */
        activeAntenna = 0, 
        /** uint8 - enum {4fps = 0 , 50fps, 150fps} */
        rfProfile = 0, 
        /** uint8 - enum {0mW = 0, 10mW, 25mW, 100mW, 500mW, 1000mW, 2000mW, 250mW, 50mW} */
        upRfPower = 0, 
        /** uint8 - Downlink RSSI (dBm * -1) */
        downRssi = 0, 
        /** uint8 - Downlink Package success rate / Link quality (%) */
        downLinkQuality = 0, 
        /** int8 - Downlink SNR (dB) */
        downSnr = 0, syncByte = SERIAL_SYNC_BYTE) {
            this.puRssiAnt1 = puRssiAnt1;
            this.puRssiAnt2 = puRssiAnt2;
            this.upLinkQuality = upLinkQuality;
            this.upSnr = upSnr;
            this.activeAntenna = activeAntenna;
            this.rfProfile = rfProfile;
            this.upRfPower = upRfPower;
            this.downRssi = downRssi;
            this.downLinkQuality = downLinkQuality;
            this.downSnr = downSnr;
            this.syncByte = syncByte;
        }
        static get payloadSize() {
            return (1 + // puRssiAnt1 uint8
                1 + // puRssiAnt2 uint8
                1 + // upLinkQuality uint8
                1 + // upSnr int8
                1 + // activeAntenna uint8
                1 + // rfProfile uint8
                1 + // upRfPower uint8
                1 + // downRssi uint8
                1 + // downLinkQuality uint8
                1 // downSnr int8
            );
        }
        get frameType() {
            return FRAME_TYPE.LINK_STATISTICS;
        }
        get crossfireFrame() {
            const payload = new Uint8Array(LinkStatistics.payloadSize);
            const payloadView = new DataView(payload.buffer);
            let offset = 0;
            payloadView.setUint8(offset, this.puRssiAnt1);
            offset += 1;
            payloadView.setUint8(offset, this.puRssiAnt2);
            offset += 1;
            payloadView.setUint8(offset, this.upLinkQuality);
            offset += 1;
            payloadView.setInt8(offset, this.upSnr);
            offset += 1;
            payloadView.setUint8(offset, this.activeAntenna);
            offset += 1;
            payloadView.setUint8(offset, this.rfProfile);
            offset += 1;
            payloadView.setUint8(offset, this.upRfPower);
            offset += 1;
            payloadView.setUint8(offset, this.downRssi);
            offset += 1;
            payloadView.setUint8(offset, this.downLinkQuality);
            offset += 1;
            payloadView.setInt8(offset, this.downSnr);
            offset += 1;
            return new CrossfireFrame(this.syncByte, this.frameType, new Uint8Array(payloadView.buffer));
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.LINK_STATISTICS) {
                throw new Error(`Invalid frame type for LinkStatistics: ${frame.type.toString(16)}`);
            }
            if (frame.payload.length < LinkStatistics.payloadSize) {
                return UnsupportedOrInvalid.fromFrame(frame);
            }
            const payloadView = new DataView(frame.payload.buffer, frame.payload.byteOffset, frame.payload.byteLength);
            let offset = 0;
            const puRssiAnt1 = payloadView.getUint8(offset);
            offset += 1;
            const puRssiAnt2 = payloadView.getUint8(offset);
            offset += 1;
            const upLinkQuality = payloadView.getUint8(offset);
            offset += 1;
            const upSnr = payloadView.getInt8(offset);
            offset += 1;
            const activeAntenna = payloadView.getUint8(offset);
            offset += 1;
            const rfProfile = payloadView.getUint8(offset);
            offset += 1;
            const upRfPower = payloadView.getUint8(offset);
            offset += 1;
            const downRssi = payloadView.getUint8(offset);
            offset += 1;
            const downLinkQuality = payloadView.getUint8(offset);
            offset += 1;
            const downSnr = payloadView.getInt8(offset);
            offset += 1;
            return new LinkStatistics(puRssiAnt1, puRssiAnt2, upLinkQuality, upSnr, activeAntenna, rfProfile, upRfPower, downRssi, downLinkQuality, downSnr, frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "LinkStatistics");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        LinkStatistics = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return LinkStatistics = _classThis;
})();

/** Converts microseconds to 11-bit ticks used in RC channels */
const usToTicks = (us) => Math.round(((us - 1500) * 8) / 5 + 992);
/** Converts 11-bit ticks used in RC channels to microseconds */
const ticksToUs = (ticks) => Math.round(((ticks - 992) * 5) / 8 + 1500);
/** Clamps a number to the 0-2047 (11-bit) range */
const clamp11$1 = (v) => Math.max(0, Math.min(0x7ff, v));
/** Packs 16 channels of 11 bits each into a 22-byte Uint8Array */
const packChannels = (channels, out) => {
    if (channels.length !== 16)
        throw new Error("channels array must have exactly 16 elements");
    let bitPos = 0;
    for (let ch = 0; ch < channels.length; ch++) {
        const value = clamp11$1(usToTicks(channels[ch]));
        for (let i = 0; i < 11; i++) {
            if (value & (1 << i)) {
                const byteIndex = (bitPos + i) >> 3;
                const bitIndex = (bitPos + i) & 7;
                out[byteIndex] |= 1 << bitIndex;
            }
        }
        bitPos += 11;
    }
};
/** Unpacks 16 channels of 11 bits each from a 22-byte Uint8Array */
const unpackChannels$1 = (payload) => {
    const channels = new Array(16).fill(1000);
    let bitPos = 0;
    for (let ch = 0; ch < 16; ch++) {
        let value = 0;
        for (let i = 0; i < 11; i++) {
            const byteIndex = (bitPos + i) >> 3;
            const bitIndex = (bitPos + i) & 7;
            if (payload[byteIndex] & (1 << bitIndex)) {
                value |= 1 << i;
            }
        }
        channels[ch] = ticksToUs(value);
        bitPos += 11;
    }
    return channels;
};
/** 16 channels packed into 22 bytes */
let RCChannelsPacked = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var RCChannelsPacked = _classThis = class {
        constructor(channel1 = 1000, channel2 = 1000, channel3 = 1000, channel4 = 1000, channel5 = 1000, channel6 = 1000, channel7 = 1000, channel8 = 1000, channel9 = 1000, channel10 = 1000, channel11 = 1000, channel12 = 1000, channel13 = 1000, channel14 = 1000, channel15 = 1000, channel16 = 1000, syncByte = SERIAL_SYNC_BYTE) {
            this.channel1 = channel1;
            this.channel2 = channel2;
            this.channel3 = channel3;
            this.channel4 = channel4;
            this.channel5 = channel5;
            this.channel6 = channel6;
            this.channel7 = channel7;
            this.channel8 = channel8;
            this.channel9 = channel9;
            this.channel10 = channel10;
            this.channel11 = channel11;
            this.channel12 = channel12;
            this.channel13 = channel13;
            this.channel14 = channel14;
            this.channel15 = channel15;
            this.channel16 = channel16;
            this.syncByte = syncByte;
        }
        static get payloadSize() {
            return 22;
        }
        get frameType() {
            return FRAME_TYPE.RC_CHANNELS_PACKED;
        }
        get crossfireFrame() {
            const payload = new Uint8Array(RCChannelsPacked.payloadSize);
            packChannels([
                this.channel1,
                this.channel2,
                this.channel3,
                this.channel4,
                this.channel5,
                this.channel6,
                this.channel7,
                this.channel8,
                this.channel9,
                this.channel10,
                this.channel11,
                this.channel12,
                this.channel13,
                this.channel14,
                this.channel15,
                this.channel16,
            ], payload);
            return new CrossfireFrame(this.syncByte, this.frameType, payload);
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.RC_CHANNELS_PACKED) {
                throw new Error(`Invalid frame type for RCChannelsPacked: ${frame.type.toString(16)}`);
            }
            if (frame.payload.length < RCChannelsPacked.payloadSize) {
                return new UnsupportedOrInvalid(frame.type, frame.payload, frame.syncByte);
            }
            const channels = unpackChannels$1(frame.payload);
            return new RCChannelsPacked(...channels, frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "RCChannelsPacked");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        RCChannelsPacked = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return RCChannelsPacked = _classThis;
})();

/** This frame is discouraged for implementation. */
let SubsetRCChannelsPacked = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var SubsetRCChannelsPacked = _classThis = class {
        constructor(buffer, syncByte = SERIAL_SYNC_BYTE) {
            this.buffer = buffer;
            this.syncByte = syncByte;
        }
        static get payloadSize() {
            return 0; // variable size
        }
        get frameType() {
            return FRAME_TYPE.SUBSET_RC_CHANNELS_PACKED;
        }
        get crossfireFrame() {
            return new CrossfireFrame(this.syncByte, this.frameType, this.buffer);
        }
        static fromFrame(frame) {
            return new SubsetRCChannelsPacked(frame.payload, frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "SubsetRCChannelsPacked");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        SubsetRCChannelsPacked = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return SubsetRCChannelsPacked = _classThis;
})();

/** Clamps a number to the 0-2047 (11-bit) range */
const clamp11 = (v) => Math.max(0, Math.min(0x7ff, v));
/** Packs 16 channels of 11 bits each into a 22-byte Uint8Array */
const pack11BitChannels = (channels, out) => {
    if (channels.length !== 16)
        throw new Error("channels array must have exactly 16 elements");
    let bitPos = 0;
    for (let ch = 0; ch < channels.length; ch++) {
        const value = clamp11(channels[ch]);
        for (let i = 0; i < 11; i++) {
            if (value & (1 << i)) {
                const byteIndex = (bitPos + i) >> 3;
                const bitIndex = (bitPos + i) & 7;
                out[byteIndex] |= 1 << bitIndex;
            }
        }
        bitPos += 11;
    }
};
/** Unpacks 16 channels of 11 bits each from a 22-byte Uint8Array */
const unpackChannels = (payload) => {
    const channels = new Array(16).fill(1024);
    let bitPos = 0;
    for (let ch = 0; ch < 16; ch++) {
        let value = 0;
        for (let i = 0; i < 11; i++) {
            const byteIndex = (bitPos + i) >> 3;
            const bitIndex = (bitPos + i) & 7;
            if (payload[byteIndex] & (1 << bitIndex)) {
                value |= 1 << i;
            }
        }
        channels[ch] = clamp11(value);
        bitPos += 11;
    }
    return channels;
};
/** 16 channels packed into 22 bytes. Channels are ticks not microseconds */
let RCChannelsPacked11Bits = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var RCChannelsPacked11Bits = _classThis = class {
        constructor(channel1Ticks = 1024, channel2Ticks = 1024, channel3Ticks = 1024, channel4Ticks = 1024, channel5Ticks = 1024, channel6Ticks = 1024, channel7Ticks = 1024, channel8Ticks = 1024, channel9Ticks = 1024, channel10Ticks = 1024, channel11Ticks = 1024, channel12Ticks = 1024, channel13Ticks = 1024, channel14Ticks = 1024, channel15Ticks = 1024, channel16Ticks = 1024, syncByte = SERIAL_SYNC_BYTE) {
            this.channel1Ticks = channel1Ticks;
            this.channel2Ticks = channel2Ticks;
            this.channel3Ticks = channel3Ticks;
            this.channel4Ticks = channel4Ticks;
            this.channel5Ticks = channel5Ticks;
            this.channel6Ticks = channel6Ticks;
            this.channel7Ticks = channel7Ticks;
            this.channel8Ticks = channel8Ticks;
            this.channel9Ticks = channel9Ticks;
            this.channel10Ticks = channel10Ticks;
            this.channel11Ticks = channel11Ticks;
            this.channel12Ticks = channel12Ticks;
            this.channel13Ticks = channel13Ticks;
            this.channel14Ticks = channel14Ticks;
            this.channel15Ticks = channel15Ticks;
            this.channel16Ticks = channel16Ticks;
            this.syncByte = syncByte;
        }
        static get payloadSize() {
            return 22;
        }
        get frameType() {
            return FRAME_TYPE.RC_CHANNELS_PACKED_11_BITS;
        }
        get crossfireFrame() {
            const payload = new Uint8Array(RCChannelsPacked11Bits.payloadSize);
            pack11BitChannels([
                this.channel1Ticks,
                this.channel2Ticks,
                this.channel3Ticks,
                this.channel4Ticks,
                this.channel5Ticks,
                this.channel6Ticks,
                this.channel7Ticks,
                this.channel8Ticks,
                this.channel9Ticks,
                this.channel10Ticks,
                this.channel11Ticks,
                this.channel12Ticks,
                this.channel13Ticks,
                this.channel14Ticks,
                this.channel15Ticks,
                this.channel16Ticks,
            ], payload);
            return new CrossfireFrame(this.syncByte, this.frameType, payload);
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.RC_CHANNELS_PACKED_11_BITS) {
                throw new Error(`Invalid frame type for RCChannelsPacked: ${frame.type.toString(16)}`);
            }
            if (frame.payload.length < RCChannelsPacked11Bits.payloadSize) {
                return new UnsupportedOrInvalid(frame.type, frame.payload, frame.syncByte);
            }
            const channels = unpackChannels(frame.payload);
            return new RCChannelsPacked11Bits(...channels, frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "RCChannelsPacked11Bits");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        RCChannelsPacked11Bits = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return RCChannelsPacked11Bits = _classThis;
})();

/** This frame is reserved for future Crossfire implementations and should not be used. */
let ReservedCrossfire0 = (() => {
    var _ReservedCrossfire0_buffer;
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var ReservedCrossfire0 = _classThis = class {
        constructor(buffer = new Uint8Array(0), syncByte = SERIAL_SYNC_BYTE) {
            this.syncByte = syncByte;
            _ReservedCrossfire0_buffer.set(this, void 0);
            __classPrivateFieldSet(this, _ReservedCrossfire0_buffer, buffer, "f");
        }
        static get payloadSize() {
            return 0; // unknown size
        }
        get frameType() {
            return FRAME_TYPE.RESERVED_CROSSFIRE_0;
        }
        get crossfireFrame() {
            return new CrossfireFrame(this.syncByte, FRAME_TYPE.RESERVED_CROSSFIRE_0, __classPrivateFieldGet(this, _ReservedCrossfire0_buffer, "f"));
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.RESERVED_CROSSFIRE_0) {
                throw new Error(`Invalid frame type for ReservedCrossfire0: ${frame.type.toString(16)}`);
            }
            return new ReservedCrossfire0(frame.payload, frame.syncByte);
        }
    };
    _ReservedCrossfire0_buffer = new WeakMap();
    __setFunctionName(_classThis, "ReservedCrossfire0");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        ReservedCrossfire0 = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return ReservedCrossfire0 = _classThis;
})();

/** This frame is reserved for future Crossfire implementations and should not be used. */
let ReservedCrossfire1 = (() => {
    var _ReservedCrossfire1_buffer;
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var ReservedCrossfire1 = _classThis = class {
        constructor(buffer = new Uint8Array(0), syncByte = SERIAL_SYNC_BYTE) {
            this.syncByte = syncByte;
            _ReservedCrossfire1_buffer.set(this, void 0);
            __classPrivateFieldSet(this, _ReservedCrossfire1_buffer, buffer, "f");
        }
        static get payloadSize() {
            return 0; // unknown size
        }
        get frameType() {
            return FRAME_TYPE.RESERVED_CROSSFIRE_1;
        }
        get crossfireFrame() {
            return new CrossfireFrame(this.syncByte, FRAME_TYPE.RESERVED_CROSSFIRE_1, __classPrivateFieldGet(this, _ReservedCrossfire1_buffer, "f"));
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.RESERVED_CROSSFIRE_1) {
                throw new Error(`Invalid frame type for ReservedCrossfire1: ${frame.type.toString(16)}`);
            }
            return new ReservedCrossfire1(frame.payload, frame.syncByte);
        }
    };
    _ReservedCrossfire1_buffer = new WeakMap();
    __setFunctionName(_classThis, "ReservedCrossfire1");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        ReservedCrossfire1 = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return ReservedCrossfire1 = _classThis;
})();

/** This frame is reserved for future Crossfire implementations and should not be used. */
let ReservedCrossfire2 = (() => {
    var _ReservedCrossfire2_buffer;
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var ReservedCrossfire2 = _classThis = class {
        constructor(buffer = new Uint8Array(0), syncByte = SERIAL_SYNC_BYTE) {
            this.syncByte = syncByte;
            _ReservedCrossfire2_buffer.set(this, void 0);
            __classPrivateFieldSet(this, _ReservedCrossfire2_buffer, buffer, "f");
        }
        static get payloadSize() {
            return 0; // unknown size
        }
        get frameType() {
            return FRAME_TYPE.RESERVED_CROSSFIRE_2;
        }
        get crossfireFrame() {
            return new CrossfireFrame(this.syncByte, FRAME_TYPE.RESERVED_CROSSFIRE_2, __classPrivateFieldGet(this, _ReservedCrossfire2_buffer, "f"));
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.RESERVED_CROSSFIRE_2) {
                throw new Error(`Invalid frame type for ReservedCrossfire2: ${frame.type.toString(16)}`);
            }
            return new ReservedCrossfire2(frame.payload, frame.syncByte);
        }
    };
    _ReservedCrossfire2_buffer = new WeakMap();
    __setFunctionName(_classThis, "ReservedCrossfire2");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        ReservedCrossfire2 = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return ReservedCrossfire2 = _classThis;
})();

let LinkStatisticsRx = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var LinkStatisticsRx = _classThis = class {
        constructor(
        /** uint8 - RSSI (dBm * -1) */
        rssiDb = 0, 
        /** uint8 - RSSI in percent */
        rssiPercent = 0, 
        /** uint8 - Package success rate / Link quality (%) */
        linkQuality = 0, 
        /** int8 - SNR (dB) */
        snr = 0, 
        /** uint8 - rf power in dBm */
        rfPowerDb = 0, syncByte = SERIAL_SYNC_BYTE) {
            this.rssiDb = rssiDb;
            this.rssiPercent = rssiPercent;
            this.linkQuality = linkQuality;
            this.snr = snr;
            this.rfPowerDb = rfPowerDb;
            this.syncByte = syncByte;
        }
        static get payloadSize() {
            return (1 + // rssiDb
                1 + // rssiPercent
                1 + // linkQuality
                1 + // snr
                1 // rfPowerDb
            );
        }
        get frameType() {
            return FRAME_TYPE.LINK_STATISTICS_RX;
        }
        get crossfireFrame() {
            const payload = new Uint8Array(LinkStatisticsRx.payloadSize);
            const payloadView = new DataView(payload.buffer);
            payloadView.setUint8(0, this.rssiDb);
            payloadView.setUint8(1, this.rssiPercent);
            payloadView.setUint8(2, this.linkQuality);
            payloadView.setInt8(3, this.snr);
            payloadView.setUint8(4, this.rfPowerDb);
            return new CrossfireFrame(this.syncByte, FRAME_TYPE.LINK_STATISTICS_RX, new Uint8Array(payloadView.buffer));
        }
        static fromFrame(frame) {
            if (frame.payload.length !== LinkStatisticsRx.payloadSize) {
                throw new Error(`Invalid payload size for LinkStatisticsRx: ${frame.payload.length}`);
            }
            if (frame.payload.length < LinkStatisticsRx.payloadSize) {
                return new UnsupportedOrInvalid(frame.type, frame.payload, frame.syncByte);
            }
            const payloadView = new DataView(frame.payload.buffer, frame.payload.byteOffset, frame.payload.byteLength);
            let offset = 0;
            const rssiDb = payloadView.getUint8(offset);
            offset += 1;
            const rssiPercent = payloadView.getUint8(offset);
            offset += 1;
            const linkQuality = payloadView.getUint8(offset);
            offset += 1;
            const snr = payloadView.getInt8(offset);
            offset += 1;
            const rfPowerDb = payloadView.getUint8(offset);
            offset += 1;
            return new LinkStatisticsRx(rssiDb, rssiPercent, linkQuality, snr, rfPowerDb, frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "LinkStatisticsRx");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        LinkStatisticsRx = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return LinkStatisticsRx = _classThis;
})();

let LinkStatisticsTx = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var LinkStatisticsTx = _classThis = class {
        constructor(
        /** uint8 - RSSI (dBm * -1) */
        rssiDb = 0, 
        /** uint8 - RSSI in percent */
        rssiPercent = 0, 
        /** uint8 - Package success rate / Link quality (%) */
        linkQuality = 0, 
        /** int8 - SNR (dB) */
        snr = 0, 
        /** uint8 - rf power in dBm */
        rfPowerDb = 0, 
        /** uint8 - rf frames per second (fps / 10) */
        fps = 0, syncByte = SERIAL_SYNC_BYTE) {
            this.rssiDb = rssiDb;
            this.rssiPercent = rssiPercent;
            this.linkQuality = linkQuality;
            this.snr = snr;
            this.rfPowerDb = rfPowerDb;
            this.fps = fps;
            this.syncByte = syncByte;
        }
        static get payloadSize() {
            return (1 + // rssiDb
                1 + // rssiPercent
                1 + // linkQuality
                1 + // snr
                1 + // rfPowerDb
                1 // fps
            );
        }
        get frameType() {
            return FRAME_TYPE.LINK_STATISTICS_TX;
        }
        get crossfireFrame() {
            const payload = new Uint8Array(LinkStatisticsTx.payloadSize);
            const payloadView = new DataView(payload.buffer);
            payloadView.setUint8(0, this.rssiDb);
            payloadView.setUint8(1, this.rssiPercent);
            payloadView.setUint8(2, this.linkQuality);
            payloadView.setInt8(3, this.snr);
            payloadView.setUint8(4, this.rfPowerDb);
            payloadView.setUint8(5, this.fps);
            return new CrossfireFrame(this.syncByte, FRAME_TYPE.LINK_STATISTICS_TX, new Uint8Array(payloadView.buffer));
        }
        static fromFrame(frame) {
            if (frame.payload.length !== LinkStatisticsTx.payloadSize) {
                throw new Error(`Invalid payload size for LinkStatisticsRx: ${frame.payload.length}`);
            }
            if (frame.payload.length < LinkStatisticsTx.payloadSize) {
                return new UnsupportedOrInvalid(frame.type, frame.payload, frame.syncByte);
            }
            const payloadView = new DataView(frame.payload.buffer, frame.payload.byteOffset, frame.payload.byteLength);
            let offset = 0;
            const rssiDb = payloadView.getUint8(offset);
            offset += 1;
            const rssiPercent = payloadView.getUint8(offset);
            offset += 1;
            const linkQuality = payloadView.getUint8(offset);
            offset += 1;
            const snr = payloadView.getInt8(offset);
            offset += 1;
            const rfPowerDb = payloadView.getUint8(offset);
            offset += 1;
            const fps = payloadView.getUint8(offset);
            offset += 1;
            return new LinkStatisticsTx(rssiDb, rssiPercent, linkQuality, snr, rfPowerDb, fps, frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "LinkStatisticsTx");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        LinkStatisticsTx = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return LinkStatisticsTx = _classThis;
})();

let Attitude = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var Attitude = _classThis = class {
        constructor(
        /** int16 - Pitch angle (LSB = 100 µrad) */
        pitch = 0, 
        /** int16 - Roll angle (LSB = 100 µrad) */
        roll = 0, 
        /** int16 - Yaw angle (LSB = 100 µrad) */
        yaw = 0, syncByte = SERIAL_SYNC_BYTE) {
            this.pitch = pitch;
            this.roll = roll;
            this.yaw = yaw;
            this.syncByte = syncByte;
        }
        static get payloadSize() {
            return (2 + // pitch int16
                2 + // roll int16
                2 // yaw int16
            );
        }
        get frameType() {
            return FRAME_TYPE.ATTITUDE;
        }
        get crossfireFrame() {
            const payload = new Uint8Array(Attitude.payloadSize);
            const payloadView = new DataView(payload.buffer);
            let offset = 0;
            payloadView.setInt16(offset, this.pitch);
            offset += 2;
            payloadView.setInt16(offset, this.roll);
            offset += 2;
            payloadView.setInt16(offset, this.yaw);
            offset += 2;
            return new CrossfireFrame(this.syncByte, this.frameType, new Uint8Array(payloadView.buffer));
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.ATTITUDE) {
                throw new Error(`Invalid frame type for Attitude: ${frame.type.toString(16)}`);
            }
            if (frame.payload.length < Attitude.payloadSize) {
                return new UnsupportedOrInvalid(frame.type, frame.payload, frame.syncByte);
            }
            const payloadView = new DataView(frame.payload.buffer, frame.payload.byteOffset, frame.payload.byteLength);
            const pitch = payloadView.getInt16(0);
            const roll = payloadView.getInt16(2);
            const yaw = payloadView.getInt16(4);
            return new Attitude(pitch, roll, yaw, frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "Attitude");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        Attitude = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return Attitude = _classThis;
})();

let MavLinkFC = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var MavLinkFC = _classThis = class {
        constructor(
        /** int16 */
        airspeed = 0, 
        /** uint8 -  vehicle mode flags, defined in MAV_MODE_FLAG enum */
        baseMode = 0, 
        /** uint32 - autopilot-specific flags */
        customMode = 0, 
        /** uint8 - FC type; defined in MAV_AUTOPILOT enum */
        autopilotType = 0, 
        /** uint8 - vehicle type; defined in MAV_TYPE enum */
        firmwareType = 0, syncByte = SERIAL_SYNC_BYTE) {
            this.airspeed = airspeed;
            this.baseMode = baseMode;
            this.customMode = customMode;
            this.autopilotType = autopilotType;
            this.firmwareType = firmwareType;
            this.syncByte = syncByte;
        }
        static get payloadSize() {
            return (2 + // airspeed int16
                1 + // baseMode uint8
                4 + // customMode uint32
                1 + // autopilotType uint8
                1 // firmwareType uint8
            );
        }
        get frameType() {
            return FRAME_TYPE.MAV_LINK_FC;
        }
        get crossfireFrame() {
            const payload = new Uint8Array(MavLinkFC.payloadSize);
            const payloadView = new DataView(payload.buffer);
            let offset = 0;
            payloadView.setInt16(offset, this.airspeed);
            offset += 2;
            payloadView.setUint8(offset, this.baseMode);
            offset += 1;
            payloadView.setUint32(offset, this.customMode);
            offset += 4;
            payloadView.setUint8(offset, this.autopilotType);
            offset += 1;
            payloadView.setUint8(offset, this.firmwareType);
            offset += 1;
            return new CrossfireFrame(this.syncByte, this.frameType, payload);
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.MAV_LINK_FC) {
                throw new Error(`Invalid frame type for MavLinkFC: ${frame.type.toString(16)}`);
            }
            if (frame.payload.length < MavLinkFC.payloadSize) {
                return new UnsupportedOrInvalid(frame.type, frame.payload, frame.syncByte);
            }
            const payloadView = new DataView(frame.payload.buffer, frame.payload.byteOffset, frame.payload.byteLength);
            let offset = 0;
            const airspeed = payloadView.getInt16(offset);
            offset += 2;
            const baseMode = payloadView.getUint8(offset);
            offset += 1;
            const customMode = payloadView.getUint32(offset);
            offset += 4;
            const autopilotType = payloadView.getUint8(offset);
            offset += 1;
            const firmwareType = payloadView.getUint8(offset);
            offset += 1;
            return new MavLinkFC(airspeed, baseMode, customMode, autopilotType, firmwareType, frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "MavLinkFC");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        MavLinkFC = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return MavLinkFC = _classThis;
})();

const MAX_MODE_STRING_LENGTH = MAX_PAYLOAD_SIZE - 1; // minus null terminator
let FlightMode = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var FlightMode = _classThis = class {
        constructor(mode, syncByte = SERIAL_SYNC_BYTE) {
            this.mode = mode;
            this.syncByte = syncByte;
            if (encodeCString(mode).length > MAX_PAYLOAD_SIZE) {
                throw new RangeError(`Flight mode string exceeds maximum length of ${MAX_MODE_STRING_LENGTH} characters`);
            }
        }
        static get payloadSize() {
            return 0; // variable
        }
        get frameType() {
            return FRAME_TYPE.FLIGHT_MODE;
        }
        get crossfireFrame() {
            const payload = encodeCString(this.mode);
            if (payload.length > MAX_PAYLOAD_SIZE) {
                throw new RangeError(`Encoded flight mode string exceeds maximum payload size of ${MAX_PAYLOAD_SIZE} bytes`);
            }
            return new CrossfireFrame(this.syncByte, this.frameType, payload);
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.FLIGHT_MODE) {
                throw new Error(`Invalid frame type for Flight Mode: ${frame.type.toString(16)}`);
            }
            return new FlightMode(decodeCString(frame.payload), frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "FlightMode");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        FlightMode = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return FlightMode = _classThis;
})();

let EspNowMessage = (() => {
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var EspNowMessage = _classThis = class {
        constructor(
        /** uint8 - Used for Seat Position of the Pilot */
        val1 = 0, 
        /** uint8 - Used for the Current Pilots Lap */
        val2 = 0, 
        /** 15 characters for the lap time current/split */
        val3 = "", 
        /** 15 characters for the lap time current/split */
        val4 = "", 
        /** Free text of 20 character at the bottom of the screen */
        freeText = "", syncByte = SERIAL_SYNC_BYTE) {
            this.val1 = val1;
            this.val2 = val2;
            this.val3 = val3;
            this.val4 = val4;
            this.freeText = freeText;
            this.syncByte = syncByte;
        }
        static get payloadSize() {
            return (1 + // val1 uint8
                1 + // val2 uint8
                15 + // val3 string
                15 + // val4 string
                20 // freeText string
            );
        }
        get frameType() {
            return FRAME_TYPE.ESP_NOW_MESSAGE;
        }
        get crossfireFrame() {
            const payload = new Uint8Array(EspNowMessage.payloadSize);
            const payloadView = new DataView(payload.buffer);
            let offset = 0;
            payloadView.setUint8(offset, this.val1);
            offset += 1;
            payloadView.setUint8(offset, this.val2);
            offset += 1;
            const val3Bytes = new TextEncoder().encode(this.val3);
            payload.set(val3Bytes.slice(0, 15), offset);
            offset += 15;
            const val4Bytes = new TextEncoder().encode(this.val4);
            payload.set(val4Bytes.slice(0, 15), offset);
            offset += 15;
            const freeTextBytes = new TextEncoder().encode(this.freeText);
            payload.set(freeTextBytes.slice(0, 20), offset);
            offset += 20;
            return new CrossfireFrame(this.syncByte, this.frameType, payload);
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.ESP_NOW_MESSAGE) {
                throw new Error(`Invalid frame type for EspNowMessage: ${frame.type.toString(16)}`);
            }
            if (frame.payload.length < EspNowMessage.payloadSize) {
                return new UnsupportedOrInvalid(frame.type, frame.payload, frame.syncByte);
            }
            const payloadView = new DataView(frame.payload.buffer, frame.payload.byteOffset, frame.payload.byteLength);
            let offset = 0;
            const val1 = payloadView.getUint8(offset);
            offset += 1;
            const val2 = payloadView.getUint8(offset);
            offset += 1;
            const val3Bytes = frame.payload.slice(offset, offset + 15);
            const val3 = new TextDecoder().decode(val3Bytes).replace(/\0.*$/g, "");
            offset += 15;
            const val4Bytes = frame.payload.slice(offset, offset + 15);
            const val4 = new TextDecoder().decode(val4Bytes).replace(/\0.*$/g, "");
            offset += 15;
            const freeTextBytes = frame.payload.slice(offset, offset + 20);
            const freeText = new TextDecoder().decode(freeTextBytes).replace(/\0.*$/g, "");
            offset += 20;
            return new EspNowMessage(val1, val2, val3, val4, freeText, frame.syncByte);
        }
    };
    __setFunctionName(_classThis, "EspNowMessage");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        EspNowMessage = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return EspNowMessage = _classThis;
})();

let Reserved = (() => {
    var _Reserved_buffer;
    let _classDecorators = [staticImplements()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    var Reserved = _classThis = class {
        constructor(buffer = new Uint8Array(0), syncByte = SERIAL_SYNC_BYTE) {
            this.syncByte = syncByte;
            _Reserved_buffer.set(this, void 0);
            __classPrivateFieldSet(this, _Reserved_buffer, buffer, "f");
        }
        static get payloadSize() {
            return 0; // unknown size
        }
        get frameType() {
            return FRAME_TYPE.RESERVED;
        }
        get crossfireFrame() {
            return new CrossfireFrame(this.syncByte, this.frameType, __classPrivateFieldGet(this, _Reserved_buffer, "f"));
        }
        static fromFrame(frame) {
            if (frame.type !== FRAME_TYPE.RESERVED) {
                throw new Error(`Invalid frame type for Reserved: ${frame.type.toString(16)}`);
            }
            return new Reserved(frame.payload, frame.syncByte);
        }
    };
    _Reserved_buffer = new WeakMap();
    __setFunctionName(_classThis, "Reserved");
    (() => {
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
        __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        Reserved = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        __runInitializers(_classThis, _classExtraInitializers);
    })();
    return Reserved = _classThis;
})();

/**
 * Get the appropriate frame variant based on the frame type.
 * The returned type can be used within switch statement to narrow down the type.
 *
 * @example
 * ```ts
 * const variant = getFrameVariant(frame);
 * switch (variant.frameType) {
 *   case FRAME_TYPE.GPS:
 *     // variant is of type GPS here
 *     break;
 * */
const getFrameVariant = (frame) => {
    switch (frame.type) {
        case FRAME_TYPE.GPS:
            return GPS.fromFrame(frame);
        case FRAME_TYPE.GPS_TIME:
            return GPSTime.fromFrame(frame);
        case FRAME_TYPE.GPS_EXTENDED:
            return GPSExtended.fromFrame(frame);
        case FRAME_TYPE.VARIOMETER_SENSOR:
            return VariometerSensor.fromFrame(frame);
        case FRAME_TYPE.BATTERY_SENSOR:
            return BatterySensor.fromFrame(frame);
        case FRAME_TYPE.BAROMETRIC_ALTITUDE_VERTICAL_SPEED:
            return BarometricAltitudeVerticalSpeed.fromFrame(frame);
        case FRAME_TYPE.AIRSPEED:
            return Airspeed.fromFrame(frame);
        case FRAME_TYPE.HEARTBEAT:
            return Heartbeat.fromFrame(frame);
        case FRAME_TYPE.RPM:
            return RMP.fromFrame(frame);
        case FRAME_TYPE.TEMP:
            return Temp.fromFrame(frame);
        case FRAME_TYPE.VOLTAGES:
            return Voltages.fromFrame(frame);
        case FRAME_TYPE.DISCONTINUED:
            return Discontinued.fromFrame(frame);
        case FRAME_TYPE.VTX_TELEMETRY:
            return VTXTelemetry.fromFrame(frame);
        case FRAME_TYPE.BAROMETER:
            return Barometer.fromFrame(frame);
        case FRAME_TYPE.MAGNETOMETER:
            return Magnetometer.fromFrame(frame);
        case FRAME_TYPE.ACCEL_GYRO:
            return AccelGyro.fromFrame(frame);
        case FRAME_TYPE.LINK_STATISTICS:
            return LinkStatistics.fromFrame(frame);
        case FRAME_TYPE.RC_CHANNELS_PACKED:
            return RCChannelsPacked.fromFrame(frame);
        case FRAME_TYPE.SUBSET_RC_CHANNELS_PACKED:
            return SubsetRCChannelsPacked.fromFrame(frame);
        case FRAME_TYPE.RC_CHANNELS_PACKED_11_BITS:
            return RCChannelsPacked11Bits.fromFrame(frame);
        case FRAME_TYPE.RESERVED_CROSSFIRE_0:
            return ReservedCrossfire0.fromFrame(frame);
        case FRAME_TYPE.RESERVED_CROSSFIRE_1:
            return ReservedCrossfire1.fromFrame(frame);
        case FRAME_TYPE.RESERVED_CROSSFIRE_2:
            return ReservedCrossfire2.fromFrame(frame);
        case FRAME_TYPE.LINK_STATISTICS_RX:
            return LinkStatisticsRx.fromFrame(frame);
        case FRAME_TYPE.LINK_STATISTICS_TX:
            return LinkStatisticsTx.fromFrame(frame);
        case FRAME_TYPE.ATTITUDE:
            return Attitude.fromFrame(frame);
        case FRAME_TYPE.MAV_LINK_FC:
            return MavLinkFC.fromFrame(frame);
        case FRAME_TYPE.FLIGHT_MODE:
            return FlightMode.fromFrame(frame);
        case FRAME_TYPE.ESP_NOW_MESSAGE:
            return EspNowMessage.fromFrame(frame);
        case FRAME_TYPE.RESERVED:
            return Reserved.fromFrame(frame);
        default:
            return UnsupportedOrInvalid.fromFrame(frame);
    }
};

// https://github.com/tbs-fpv/tbs-crsf-spec/blob/main/crsf.md
// ardupilot/libraries/AP_RCProtocol/AP_RCProtocol_CRSF.h
class CrossfireParser {
    constructor(onFrame) {
        this.onFrame = onFrame;
        this.buffer = new Uint8Array(0);
        this.offset = 0;
    }
    appendChunk(data) {
        // hard sanity guard
        if (this.offset > this.buffer.length) {
            // parser invariant broken → reset safely
            this.buffer = new Uint8Array(0);
            this.offset = 0;
        }
        const newBuffer = data instanceof ArrayBuffer
            ? new Uint8Array(data)
            : new Uint8Array(data.buffer, data.byteOffset, data.byteLength);
        if (this.offset === this.buffer.length) {
            // previous buffer fully consumed → replace
            this.buffer = newBuffer;
        }
        else {
            // some leftover → merge
            const leftover = this.buffer.subarray(this.offset);
            const merged = new Uint8Array(leftover.length + newBuffer.length);
            merged.set(leftover, 0);
            merged.set(newBuffer, leftover.length);
            this.buffer = merged;
        }
        this.offset = 0;
        this.parseFrames();
    }
    parseFrames() {
        while (this.offset < this.buffer.length) {
            const syncOffset = this.offset;
            const syncByte = this.buffer[this.offset];
            if (!SYNC_BYTE_VARIANTS.has(syncByte)) {
                this.offset++;
                continue;
            }
            const frameLengthOffset = syncOffset + 1;
            if (frameLengthOffset >= this.buffer.length)
                break; // not enough data to read length
            const frameLength = this.buffer[frameLengthOffset];
            if (frameLength < MIN_FRAME_SIZE || frameLength > MAX_FRAME_SIZE) {
                // invalid length
                this.offset++;
                continue;
            }
            const frameTypeOffset = frameLengthOffset + 1;
            const payloadStartOffset = frameTypeOffset + 1;
            const crcOffset = payloadStartOffset + (frameLength - 2);
            if (crcOffset >= this.buffer.length)
                break; // not enough data to read full frame
            const crc = this.buffer[crcOffset];
            const crcData = this.buffer.subarray(frameTypeOffset, crcOffset); // end exclusive
            const computedCrc = getCRC(crcData);
            if (crc !== computedCrc) {
                // invalid crc, skip this sync byte
                this.offset += 1;
                continue;
            }
            const payload = this.buffer.subarray(payloadStartOffset, crcOffset); // end exclusive
            const frameType = this.buffer[frameTypeOffset];
            this.onFrame(new CrossfireFrame(syncByte, frameType, payload));
            this.offset = crcOffset + 1;
        }
    }
}

const serialize = (frame) => {
    const buffer = new Uint8Array(2 + frame.length); // sync + length + type+payload+crc
    let offset = 0;
    buffer[offset++] = frame.syncByte;
    buffer[offset++] = frame.length;
    buffer[offset++] = frame.type;
    buffer.set(frame.payload, offset);
    offset += frame.payload.length;
    buffer[offset++] = frame.crc;
    return buffer;
};

export { AccelGyro, Airspeed, Attitude, Barometer, BarometricAltitudeVerticalSpeed, BatterySensor, CrossfireFrame, CrossfireParser, DEVICE_ADDRESS, Discontinued, EspNowMessage, FRAME_TYPE, FlightMode, GPS, GPSExtended, GPSTime, Heartbeat, LinkStatistics, LinkStatisticsRx, LinkStatisticsTx, MAX_FRAME_SIZE, MAX_PAYLOAD_SIZE, MIN_FRAME_SIZE, Magnetometer, MavLinkFC, RCChannelsPacked, RCChannelsPacked11Bits, RMP, Reserved, ReservedCrossfire0, ReservedCrossfire1, ReservedCrossfire2, SERIAL_SYNC_BYTE, SYNC_BYTE_VARIANTS, SubsetRCChannelsPacked, Temp, UnsupportedOrInvalid, VTXTelemetry, VariometerSensor, Voltages, getFrameVariant, serialize, staticImplements };
